File: backend/models/session.py
from datetime import datetime, timezone
from typing import Literal
from uuid import UUID, uuid4

from beanie import Document
from pydantic import BaseModel, Field
from pymongo import IndexModel

SessionStatus = Literal["lobby", "swiping", "results", "matched"]


class SessionMember(BaseModel):
    user_id: UUID
    name: str
    joined_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class Session(Document):
    id: UUID = Field(default_factory=uuid4)
    code: str
    host_user_id: UUID
    status: SessionStatus = "lobby"
    location_lat: float | None = None
    location_lng: float | None = None
    location_label: str | None = None
    members: list[SessionMember] = Field(default_factory=list)
    matched_restaurant_id: UUID | None = None
    solo_mode: bool = False
    cuisine_overrides: list[str] | None = None
    radius_km_override: float | None = None
    budget_overrides: list[str] | None = None  # multi-select; e.g. ["$$", "$$$"] → only those tiers
    swipe_ceiling_override: int | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "sessions"
        indexes = [IndexModel("code", unique=True)]

File: mobile-swift/DishMatch/Models/Session.swift
import Foundation

enum SessionStatus: String, Codable, Equatable {
    case lobby, swiping, results, matched
}

struct Session: Codable, Identifiable, Equatable {
    let id: UUID
    let code: String
    let hostUserId: UUID
    var status: SessionStatus
    let locationLat: Double?
    let locationLng: Double?
    let locationLabel: String?
    var members: [SessionMember]
    let matchedRestaurantId: UUID?
    var soloMode: Bool?
    let cuisineOverrides: [String]?
    let radiusKmOverride: Double?
    let budgetOverrides: [String]?
    let swipeCeilingOverride: Int?
    let topN: Int?
    let createdAt: Date
}

struct SessionMember: Codable, Identifiable, Equatable {
    let userId: UUID
    let name: String
    let joinedAt: Date

    var id: UUID { userId }
}
